### ISSUE TITLE
Empty field deserializes to string as "~"

### ISSUE THREAD

**Issue by dtolnay** [dtolnay/serde-yaml#87]
In the following example, both inputs deserialize to `field: "~"`. I would expect the first one to deserialize to `field: ""`.

```rust
#[macro_use]
extern crate serde_derive;

extern crate serde;
extern crate serde_yaml;

#[derive(Deserialize, Debug)]
struct S {
    field: String,
}

fn main() {
    println!("{:?}", serde_yaml::from_str::<S>("field:"));
    println!("{:?}", serde_yaml::from_str::<S>("field: ~"));
}
```

**Comment by mbr** [comment]
I ran into this as well. Maybe this could also be a deserialization error, as ~ would only be valid for an `Option<String>`?

**Comment by rtyler** [comment]
I have been doing some debugging on this trying figure out where the source of the problem is.

Using this test case:

```rust
fn test_empty_field() {
    #[derive(Deserialize, Debug)]
    struct S {
        field: String,
    }

    let yaml =
"
---
field:
";

    let thing: S = serde_yaml::from_str::<S>(&yaml).unwrap();
    assert_eq!(thing.field, "");
}
```

I've tracked things down into the [empty_scalar](https://github.com/chyh1990/yaml-rust/blob/master/src/parser.rs#L55) function in yaml-rust, which is  where the tilde comes from.

I am not sure if the bug is in serde-yaml or in yaml-rust however. Using yaml-rust just by itself:


```rust
    extern crate yaml_rust;
    use yaml_rust::{YamlLoader, YamlEmitter};

    let yaml =
"
---
field:
";
    let docs = YamlLoader::load_from_str(yaml).unwrap();
    let doc = &docs[0];
    println!("Field: {}", doc["field"].as_str().unwrap());
```

The above will blow up because we cannot call an `unwrap()` on a `None` value, which smells to me like yaml-rust is emitting the right thing, and somewhere in serde-yaml we're taking the inappropriate member of `Event::Scalar`

**Comment by eliaslevy** [comment]
Running into this issue as well when deserializing a struct using derive on key-value fields with a type of `String` when the YAML has an empty value.

**Comment by kstrohbeck** [comment]
Deserializing to `serde_yaml::Value` catches the null. I think the reason that it doesn't work for a `String` type is that `Deserializer::deserialize_string/str` forwards to `visitor.visit_str`, where pretty much all other scalar-possible values forward to `deserialize_scalar` (and then `visit_untagged_str`,) which handles the different untagged representations of scalar, including "~". It seems like changing `deserialize_str` to use `visit_untagged_str` would work, although in this case I think it would just cause the deserialization to fail. Could create a `Scalar` enum with the proper types, and then turn it back into a string representation if a string is requested?

**Comment by kstrohbeck** [comment]
I've gotten the empty field to work by replacing the `deserialize_str` match on `Event::Scalar` with one that checks for "~" and, if so, calls `visit_str("")` instead of `visit_str(v)`. It's hacky, though - it will turn a literal "~" into an empty string when deserialized, since there's no way to tell the two apart. Given that the yaml standard seems to consider null and an empty node as separate entities, it might be worth making a change to `yaml_rust` to represent an empty node separately (maybe just as a scalar with an empty string representation.)

**Comment by dtolnay** [comment]
This is fixed in 0.9.

```rust
use serde::Deserialize;

#[derive(Deserialize, Debug)]
struct S {
    field: String,
}

fn main() -> Result<(), serde_yaml::Error> {
    println!("{:?}", serde_yaml::from_str::<S>("field:"));
    println!("{:?}", serde_yaml::from_str::<S>("field: ~"));
    Ok(())
}
```

```console
Ok(S { field: "" })
Ok(S { field: "~" })
```
